### DISCUSSION TITLE
useState with computed ref

### DISCUSSION THREAD

**Discussion by bogdan0083** [nuxt/nuxt#26571] [Questions] [answered]
Hey guys! I'm trying to pass computed ref for `useState` and watch those changes in `useAsyncData` but it doesn't work.

Let me first show you **an example that actually works (without useState)**:

```vue
<script setup lang="ts">
import { ArticlesApi, type GetArticlesRequest } from '~/lib/api/__generated__'

const ITEMS_PER_PAGE = 10
const router = useRouter()
const route = useRoute()
const computedTag = computed(() => route.query.tag || 'none') as ComputedRef<string>

const currentPage = computed(() => route.query.page ?? 1)

const offset = computed(() => (currentPage.value - 1) * ITEMS_PER_PAGE)

const { data, pending, error } = await useAsyncData('articles', () => {
  const api = new ArticlesApi()
  // this works! we watch offset and computedTag refs and call api when they're updated.
  return api.getArticles({ offset: offset.value, tag: computedTag.value !== 'none' ? computedTag.value : undefined })
}, { watch: [offset, computedTag] })
```

When route query is updated (like this for example `$router.replace({ query: { page: 1, tag: 'hello' } })`) our computed are also updated and watched inside `useAsyncData` callback.

Again, the above example works just fine. But **I want to abstract `useAsyncData` call to a composable `useArticlesQuery`**:

```ts
import { ArticlesApi, type GetArticlesRequest } from '~/lib/api/__generated__'

export async function useArticlesQuery(opts: GetArticlesRequest) {
  const api = new ArticlesApi()
  
  // want to watch opts.offset and opts.tag to execute api call when those values are changed
  const response = await useAsyncData('articles', () => api.getArticles(opts), { watch: [opts.offset, opts.tag] })
  return response
}
``` 

Because I use composable I need to use `useState` to work with ssr:

```ts
const computedTag = computed(() => route.query.tag || 'none') as ComputedRef<string>

// use useState on computed ref `computedTag`
const computedTagState = useState('computedTagState', () => computedTag)

const currentPage = computed(() => route.query.page ?? 1)

// use useState on computed ref `currentPage`
const currentPageState = useState(() => currentPage)
```

but when `router` changes query params I can't watch for changes:

```ts
watchEffect(() => {
  // when router query params change this DOESN'T CHANGE. Can't watch on this computed state
  console.log(computedTagState.value)
  // when router query params change this DOESN'T CHANGE. Can't watch on this computed state
  console.log(currentPageState.value)
})
```

How do I make `useState` a computed value that properly updates when `route.query` changes? Those values will be sent to `useArticlesQuery` and watched.

**Comment by Mini-ghost** [comment]
According to Nuxt's implementation, `useState` does not automatically invoke the initialization function in every scenario. Specifically, when `useState` is first called on the client with a distinct `key`, Nuxt evaluates whether the `key`'s value already exists within `payload.state`. If the value for `payload.state[key]` is not defined, Nuxt then proceeds to utilize the initialization function to secure a default value.

In the scenario you've described, `useState` gets initialized server-side. Consequently, upon transitioning to the client side, it retrieves data directly from `payload.state` without invoking the provided initialization function. This leads to a situation where, despite alterations in `route.query.page`, the `currentPageState` remains unchanged, resulting in a misalignment between the expected and actual states.

**Reply by bogdan0083** [reply]
Thanks for the explanation. Now it's much clear for me how it works under the hood. How do I make it execute initialization function (or make useState work as a computed ref like `currentPage`) ?  Will `currentPageState.value = currentPage` work? Gonna try this workaround today.

**Reply by bogdan0083** [reply]
Tried to set `currentPageState.value` explicitly:

```ts
// use useState on computed ref `computedTag`
const computedTagState = useState('computedTagState', () => computedTag)
computedTagState.value = computedTag // setting value right away

const currentPage = computed(() => route.query.page ?? 1)

// use useState on computed ref `currentPage`
const currentPageState = useState('currentPageState', () => currentPage)
currentPageState.value = currentPage  // setting value right away

const offset = computed(() => (currentPage.value - 1) * ITEMS_PER_PAGE)

const { data, pending, error } = useArticlesQuery({ tag: computedTagState })
```

`useArticlesQuery.ts`:

```ts
import { ArticlesApi, type GetArticlesRequest } from '~/lib/api/__generated__'

export async function useArticlesQuery(opts: GetArticlesRequest) {
  const api = new ArticlesApi()

  const response = await useAsyncData('articles', () => api.getArticles(opts), {
    watch: [opts.tag],
  })
  return response
}
```

But loading page fails with:

```json
{
  "statusCode": 404,
  "statusMessage": "Cannot find any path matching /.",
  "stack": []
}
```
Using just `useArticlesQuery({})` without reactive property runs just fine.

😞

**Reply by bogdan0083** [reply, marked as answer]
I finally found a solution. I need to unwrap passed refs before it goes to `api.getArticles` (I use `toValue` for unwraping but using `reactive(opts)` would also work):

 ```ts
 import { ArticlesApi, type GetArticlesRequest } from '~/lib/api/__generated__'

interface UseArticlesQueryOptions {
  offset?: Ref<GetArticlesRequest['offset']>
  author?: Ref<GetArticlesRequest['author']>
  tag?: Ref<GetArticlesRequest['tag']>
}

export async function useArticlesQuery(opts: UseArticlesQueryOptions) {
  const api = new ArticlesApi()
  const watchSources = []

  if (opts.offset)
    watchSources.push(opts.offset)

  if (opts.tag)
    watchSources.push(opts.tag)

  const response = useAsyncData('articles', () => api.getArticles({
    offset: toValue(opts.offset),
    tag: opts.tag?.value === 'none' ? undefined : toValue(opts.tag),
  }), { watch: watchSources })

  return response
}
 ```
 
 You can pass reactive variables to a composable but we should not forget to unwrap them before it goes as an argument for api client.
